### PR TITLE
Add step_percentile

### LINKED ISSUE #765: Request `step_percentile()` , non-obvious opportunity for data leak with `step_mutate()`
## Feature

In the tidymodels developer [section ](https://www.tidymodels.org/learn/develop/recipes/), there is an example that creates a `step_percentile()` function. I think this would be great to have included in `receipes`, there is a workaround using `step_mutate()` but there is easy opportunity for data-leak that isn't entirely obvious, and I would imagine it's a relatively common transformation regardless. Any thoughts?

 ``` r
library(tidymodels)
#> Registered S3 method overwritten by 'tune':
#>   method                   from   
#>   required_pkgs.model_spec parsnip

data("car_prices")

set.seed(24)

splits <- initial_split(car_prices)
car_train <- training(splits)
car_test <- testing(splits)

car_rec_no_leak <- recipe(Price ~ . , data = car_train) %>%
  step_mutate(Mileage = ecdf(car_train$Mileage)(Mileage)) %>%
  prep()

car_rec_leak <- recipe(Price ~ . , data = car_train) %>%
  step_mutate(Mileage = ecdf(Mileage)(Mileage)) %>%
  prep()


bake(car_rec_no_leak, new_data = car_test)
#> # A tibble: 201 x 18
#>    Mileage Cylinder Doors Cruise Sound Leather Buick Cadillac Chevy Pontiac
#>      <dbl>    <int> <int>  <int> <int>   <int> <int>    <int> <int>   <int>
#>  1   0.682        4     2      1     0       0     0        0     0       0
#>  2   0.153        4     4      1     1       0     0        0     0       0
#>  3   0.164        4     4      1     1       0     0        0     0       0
#>  4   0.597        4     4      1     1       0     0        0     0       0
#>  5   0.891        4     4      1     1       1     0        0     0       0
#>  6   0.257        4     4      1     1       1     0        0     0       0
#>  7   0.597        4     4      1     1       1     0        0     0       0
#>  8   0.461        4     4      1     0       0     0        0     0       1
#>  9   0.365        4     4      1     1       0     0        0     0       1
#> 10   0.935        8     2      1     1       1     0        0     1       0
#> # ... with 191 more rows, and 8 more variables: Saab <int>, Saturn <int>,
#> #   convertible <int>, coupe <int>, hatchback <int>, sedan <int>, wagon <int>,
#> #   Price <dbl>

# Testing data is used to compute the percentile
bake(car_rec_leak, new_data = car_test)
#> # A tibble: 201 x 18
#>    Mileage Cylinder Doors Cruise Sound Leather Buick Cadillac Chevy Pontiac
#>      <dbl>    <int> <int>  <int> <int>   <int> <int>    <int> <int>   <int>
#>  1   0.652        4     2      1     0       0     0        0     0       0
#>  2   0.169        4     4      1     1       0     0        0     0       0
#>  3   0.179        4     4      1     1       0     0        0     0       0
#>  4   0.602        4     4      1     1       0     0        0     0       0
#>  5   0.925        4     4      1     1       1     0        0     0       0
#>  6   0.289        4     4      1     1       1     0        0     0       0
#>  7   0.597        4     4      1     1       1     0        0     0       0
#>  8   0.443        4     4      1     0       0     0        0     0       1
#>  9   0.333        4     4      1     1       0     0        0     0       1
#> 10   0.970        8     2      1     1       1     0        0     1       0
#> # ... with 191 more rows, and 8 more variables: Saab <int>, Saturn <int>,
#> #   convertible <int>, coupe <int>, hatchback <int>, sedan <int>, wagon <int>,
#> #   Price <dbl>
```

<sup>Created on 2021-08-05 by the [reprex package](https://reprex.tidyverse.org) (v2.0.0)</sup>

### PR DESCRIPTION
**Pull request by EmilHvitfeldt**
This PR aims to close #765 by adding `step_percentile()`. The code is pretty much a copy from https://www.tidymodels.org/learn/develop/recipes/ with the changes localized in this commit https://github.com/tidymodels/recipes/commit/ba42ad5477d3d128b5ea1ea6c546787293414a59

### CODE REVIEW

**Comment by juliasilge** [comment]
If we add this as a new step, then we'll need to rewrite the tidymodels.org article with some new example transformation; it doesn't make much sense IMO to have the article for an existing step that people can use already.

**Comment by EmilHvitfeldt** [comment]
how are we going to handle this long term? If the step we showcase in tidymodels.org is too useful will we end up getting requests for it, but it also needs to be useful enough to showcase. 

Could the example be a step which is a subset of an existing step? e.i. "this is how you can create a step that does a 2-degree polynomial expansion. If you need a general polynomial step you can use `step_poly()`". (I'm not saying `step_poly()` would be the right step to use here)

File: tests/testthat/test_percentile.R [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+    biomass_tr_baked$carbon
+  )
+  expect_equal(
+    approx(carbon_quantiles, y = 0:100, xout = biomass_te$carbon)$y/100,
+    biomass_te_baked$carbon
+  )
+  sulfur_values <- as.numeric(gsub("%$", "", names(sulfur_quantiles)))
+  expect_equal(
+    approx(sulfur_quantiles, y = sulfur_values, xout = biomass_tr$sulfur)$y/100,
+    biomass_tr_baked$sulfur
```
Comment (topepo) [inline tests/testthat/test_percentile.R:42]: Can you add a test case where there are fewer unique values in the column being transformed than percentiles requested? Maybe make a new variable that is something like `round(biomass$carbon, -1)`.

Comment (EmilHvitfeldt) [reply tests/testthat/test_percentile.R:42]: Added in https://github.com/tidymodels/recipes/pull/882/commits/e204adbe4bbcc688d7c78f6001473d6f57647842

**Pull request merged**